lib/websocket/frame.js

```diff
@@ -0,0 +1,223 @@
+      code = data.readUInt16BE(0)
+    }
+
+    if (onlyCode) {
+      if (!isValidStatusCode(code)) {
+        return null
+      }
+
+      return { code }
+    }
+
+    // https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.6
+    /** @type {Buffer} */
+    let reason = data.subarray(2)
+
+    // Remove BOM
+    if (reason[0] === 0xEF && reason[1] === 0xBB && reason[2] === 0xBF) {
+      reason = reason.subarray(3)
+    }
+
+    if (code !== undefined && !isValidStatusCode(code)) {
+      return null
+    }
+
+    try {
+      // TODO: optimize this
+      reason = new TextDecoder('utf-8', { fatal: true }).decode(reason)
+    } catch {
+      return null
+    }
+
+    return { code, reason }
+  }
+
+  /**
+   * @param {Buffer} buffer
+   */
+  static from (buffer) {
+    const fin = (buffer[0] & 0x80) !== 0
+    const rsv1 = (buffer[0] & 0x40) !== 0
+    const rsv2 = (buffer[0] & 0x20) !== 0
+    const rsv3 = (buffer[0] & 0x10) !== 0
+    const opcode = buffer[0] & 0x0F
+    const masked = (buffer[1] & 0x80) !== 0
+
+    // Data sent from an endpoint cannot be masked.
+    assert(!masked)
+
+    let payloadLength = buffer[1] & 0x7F
+    let data
+
+    if (payloadLength <= 125) {
+      data = buffer.subarray(2)
+    } else if (payloadLength === 126) {
+      payloadLength = buffer.subarray(2, 4).readUInt16BE(0)
+      data = buffer.subarray(4)
+    } else if (payloadLength === 127) {
+      payloadLength = buffer.subarray(2, 10).readBigUint64BE(0)
+      data = buffer.subarray(10)
+    }
+
+    const frame = new WebsocketFrame({
+      fin,
+      rsv1,
+      rsv2,
+      rsv3,
+      opcode,
+      payloadLength,
+      data
+    })
+
+    return frame
+  }
+}
+
+class WebsocketFrameSend {
+  /**
+   * @param {Buffer|undefined} data
+   */
+  constructor (data) {
+    this.frameData = data
+    this.maskKey = randomBytes(4)
+  }
+
+  createFrame (opcode) {
+    const bodyLength = this.frameData?.byteLength ?? 0
+
+    /** @type {number} */
+    let payloadLength = bodyLength // 0-125
+    let offset = 6
+
+    if (bodyLength > maxUnsigned16Bit) {
+      offset += 8 // payload length is next 8 bytes
+      payloadLength = 127
+    } else if (bodyLength > 125) {
+      offset += 2 // payload length is next 2 bytes
+      payloadLength = 126
+    }
+
+    const buffer = Buffer.alloc(bodyLength + offset)
+
+    buffer[0] |= 0x80 // FIN
+    buffer[0] = (buffer[0] & 0xF0) + opcode // opcode
+
+    /*! ws. MIT License. Einar Otto Stangvik <einaros@gmail.com> */
+    buffer[offset - 4] = this.maskKey[0]
+    buffer[offset - 3] = this.maskKey[1]
+    buffer[offset - 2] = this.maskKey[2]
+    buffer[offset - 1] = this.maskKey[3]
+
+    buffer[1] = payloadLength
+
+    if (payloadLength === 126) {
+      buffer.writeUInt16BE(bodyLength, 2)
+    } else if (payloadLength === 127) {
+      buffer.writeUIntBE(bodyLength, 4, 6)
+    }
+
+    buffer[1] |= 0x80 // MASK
+
+    // mask body
+    for (let i = 0; i < this.frameData?.length ?? 0; i++) {
```

Comment:
If mask is 0 then skip all of this. It's a common optimisation to set mask to 0 for this reason.

Reply:
Setting mask to 0 makes it a continuation frame, basically sending a fragmented message. We currently don't support support sending fragmented messages, and even if we did, we would still need to mask the body.

Reply:
@ronag is suggesting skip masking if the mask is `00000000`.